lib/ruby_smb/gss/provider/ntlm.rb

```diff
@@ -0,0 +1,337 @@
+                  ], 0, :CONTEXT_SPECIFIC)
+                ])
+              ], 1, :CONTEXT_SPECIFIC).to_der
+
+              account = @provider.get_account(
+                type3_msg.user,
+                domain: type3_msg.domain
+              )
+              if account.nil?
+                if @provider.allow_anonymous
+                  identity = IDENTITY_ANONYMOUS
+                end
+              else
+                identity = account.to_s
+              end
+            end
+
+            Result.new(buffer, nt_status, identity)
+          end
+        end
+
+        # @param [Boolean] allow_anonymous whether or not to allow anonymous authentication attempts
+        # @param [String] default_domain the default domain to use for authentication, unless specified 'WORKGROUP' will
+        #   be used
+        def initialize(allow_anonymous: false, default_domain: nil)
+          @allow_anonymous = allow_anonymous
+          @default_domain = default_domain || 'WORKGROUP'
+          @accounts = []
+          @generate_server_challenge = -> { SecureRandom.bytes(8) }
+
+          @dns_domain = @netbios_domain = 'LOCALDOMAIN'
+          @dns_hostname = @netbios_hostname = 'LOCALHOST'
+        end
+
+        #
+        # Generate the 8-byte server challenge. If a block is specified, it's used as the challenge generation routine
+        # and should return an 8-byte value.
+        #
+        # @return [String] an 8-byte challenge value
+        def generate_server_challenge(&block)
+          if block.nil?
+            @generate_server_challenge.call
+          else
+            @generate_server_challenge = block
+          end
+        end
+
+        def new_authenticator(server_client)
+          # build and return an instance that can process and track stateful information for a particular connection but
+          # that's backed by this particular provider
+          Authenticator.new(self, server_client)
+        end
+
+        #
+        # Lookup and return an account based on the username and optionally, the domain. If no domain is specified or
+        # or it is the special value '.', the default domain will be used. The username and domain values are case
+        # insensitive.
+        #
+        # @param [String] username the username of the account to fetch.
+        # @param [String, nil] domain the domain in which the account to fetch exists.
+        # @return [Account, nil] the account if it was found
+        def get_account(username, domain: nil)
+          # the username and password values should use the native encoding for the comparison in the #find operation
+          username = username.downcase
+          domain = @default_domain if domain.nil? || domain == '.'.encode(domain.encoding)
+          domain = domain.downcase
+          @accounts.find { |account| account.username.encode(username.encoding).downcase == username && account.domain.encode(domain.encoding).downcase == domain }
```

Comment:
Wouldn't the computer name be the authentication domain though? Say you have two non-domain joined computers both with users "bob" on them. Those are two separate users with two separate passwords that are valid in two separate contexts. It's quite feasible that bob from system A is a user we want to authenticate while bob from system B is not.

I think in this case, it's working as intended right? That's why when the user wants to authenticate, they should specify the `WORKGROUP` or `.` domain because their local computer account probably isn't valid.

If we really wanted though, maybe an option to just ignore the domain could be added to the provider.

Reply:
If I understand your example correctly, the two users belong to two separate hosts, since they are local users (non-domain users). So, there will be two separate SMB servers in this case. Each one will authenticate the user locally in its own context. I think that any authentication of a user without domain is considered local, and the domain should be ignored.

I tested this with the `net use` command:
```
C:\Users\TestUserDA>net use \\172.31.41.227\share /user:foo\administrator
The password or user name is invalid for \\172.31.41.227\share.

Enter the password for 'foo\administrator' to connect to '172.31.41.227':
The command completed successfully.
```

It successfully logged on 172.31.41.227, which is on a different domain, with the local administrator. The `NTLNSSP` blob in the `Session Setup` request contains `foo` as domain name:

![Screenshot 2021-08-18 at 21 13 13](https://user-images.githubusercontent.com/56716719/129958285-e1a30b3d-0c3b-4b87-98d3-a87d5559c698.png)

It looks like the domain name is ignored.

That said, I'm not sure if this applies in the context of this SMB server.

Reply:
> So, there will be two separate SMB servers in this case. Each one will authenticate the user locally in its own context. 

There doesn't need to be two separate servers though. If I add accounts to my one server instance with two different domains:

```
ntlm_provider = RubySMB::Gss::Provider::NTLM.new
ntlm_provider.put_account('bob', 'password1', domain: 'HOST1')
ntlm_provider.put_account('bob', 'password2', domain: 'HOST2')
```

Then bob from HOST1 and bob from HOST2 will both be able to authenticate to me given that their systems will set the domain field to their unique hostnames. This then allows me to know the identity of each and differentiate HOST1\bob from HOST2\bob.

If we ignored the domain in the lookup, HOST1\bob and HOST2\bob would need to be the same account and share the same password which seems unnecessary. If we're mostly expecting HOST1\bob to connect to us, then we could set the default domain to HOST1, and that would just mean that HOST2\bob would need to specify their hostname as the domain field which you mentioned was the behavior anyways.

Reply:
Absolutely, it makes sense this way. I didn't think of using the server to host the same local account name from different hosts, which requires the hostname to differentiate them. We will just need to know the correct hostname in advance, which I believe is is not a big deal.

Reply:
I think we want to leave this exposed to API and even user configuration as much as possible since the semantics of protocol access patterns inform defensive analysis - the more entropy we feed into whatever we put on the wire, the less chance of a fingerprint or heuristic match being built around our interactions with target systems.